src/client/src/offset.rs

```diff
@@ -0,0 +1,273 @@
+/// events at the `Absolute` offset of 0. An absolute offset is a
+/// unique index that represents the event's distance from the very
+/// beginning of the partition. The absolute offset of an event never
+/// changes.
+///
+/// Sometimes when a partition gets very large, Fluvio will begin
+/// deleting events from the beginning of the log in order to save
+/// space. Whenever it does this, it keeps track of the latest
+/// non-deleted event. This allows the `FromBeginning` offset to
+/// select an event which is a certain number of places in front of
+/// the deleted range. For example, let's say that the first two
+/// events from our partition were deleted. Our new offsets would
+/// look like this:
+///
+/// ```text
+///                          These events were deleted!
+///                          |
+///                          vvvvvv
+///         Partition Log: [ .., .., 22, 33, 44, 55, 66 ]
+///       Absolute Offset:    0,  1,  2,  3,  4,  5,  6
+///  FromBeginning Offset:            0,  1,  2,  3,  4
+///        FromEnd Offset:    6,  5,  4,  3,  2,  1,  0
+/// ```
+///
+/// Just like the `FromBeginning` offset may change if events are deleted,
+/// the `FromEnd` offset will change when new events are added. Let's take
+/// a look:
+///
+/// ```text
+///                                        These events were added!
+///                                                               |
+///                                                      vvvvvvvvvv
+///         Partition Log: [ .., .., 22, 33, 44, 55, 66, 77, 88, 99 ]
+///       Absolute Offset:    0,  1,  2,  3,  4,  5,  6,  7,  8,  9
+///  FromBeginning Offset:            0,  1,  2,  3,  4,  5,  6,  7
+///        FromEnd Offset:    9,  8,  7,  6,  5,  4,  3,  2,  1,  0
+/// ```
+///
+/// # Example
+///
+/// All offsets must be constructed with a positive index. Negative
+/// numbers are meaningless for offsets and therefore are not allowed.
+/// Trying to construct an offset with a negative number will yield `None`.
+///
+/// ```
+/// use fluvio::Offset;
+/// let absolute_offset = Offset::absolute(5).unwrap();
+/// let offset_from_beginning = Offset::from_beginning(100).unwrap();
+/// let offset_from_end = Offset::from_end(10).unwrap();
+///
+/// // Negative offsets will give None
+/// assert_eq!(Offset::absolute(-10), None);
+/// assert_eq!(Offset::from_beginning(-15), None);
+/// assert_eq!(Offset::from_end(-20), None);
+/// ```
+#[derive(Debug, Clone, PartialEq)]
+pub struct Offset {
+    inner: OffsetInner,
+}
+
+impl Offset {
+    /// Creates an absolute offset with the given index
+    ///
+    /// The index must not be less than zero.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// # use fluvio::Offset;
+    /// assert!(Offset::absolute(100).is_some());
+    /// assert!(Offset::absolute(0).is_some());
+    /// assert!(Offset::absolute(-10).is_none());
+    /// ```
+    pub fn absolute(index: i64) -> Option<Offset> {
```

Comment:
shouldn't this be u64?,   and this return option?

Reply:
@sehz I thought we talked about this... This is the same representation that `FetchOffset` was using:
https://github.com/infinyon/fluvio/pull/292/files#diff-3ec5e00be2029042b2e036bb803ac947L45-L52
I thought the purpose was to maintain consistency with a potential Java client representation of `long`. I will happily make this into `u64` if you want me to though.

Reply:
internal should be kept in the i64 to preserve Java compatibility.  The parameter should be u64 since offset should be positive.

Reply:
Well that leaves us with the same kind of problem, which is that there will be a portion of the range that is invalid. E.g. even if I wrote it as `fn absolute(index: u64)`, then sure they couldn't give a negative number, but they _could_ do something like `Offset::absolute(u64::MAX)` and cause a problem. I think that if we're going to store this internally as i64, then it'd communicate the allowed input values much better to ask for an `i64` argument but give the specification that it must not be less than zero. This also prevents any subtle mistakes with casting between integer types

Reply:
If we accept i64 then it should return error if number is negative

Reply:
Do you mean that instead of returning an Option it should return a Result? I'd be open to doing that.

Reply:
yes, invalid configuration should return Error